### PR TITLE
[SPARK-34114][SQL] should not trim right for read-side char length check and padding

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

On the read-side, we should respect the original data instead of trimming it first.

It brings extra overhead on the code-gen code side, trimming and padding for the same field, and it's also unnecessary and a bug

### Why are the changes needed?

bugfix and perf regression

### Does this PR introduce _any_ user-facing change?

no
### How was this patch tested?

new tests

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @maropu @HyukjinKwon @dongjoon-hyun, thanks for review

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala [hunk 1]
```diff
@@ -208,53 +209,56 @@ object CharVarcharUtils extends Logging {
...
-      }
+          raiseError("char", length),
+          trimmed)
+
+      case VarcharType(length) =>
+        val trimmed = if (needTrim) StringTrimRight(expr) else expr
+        // Trailing spaces do not count in the length check. We need to retain the trailing spaces
+        // (truncate to length N), as there is no read-time padding for varchar type.
+        // TODO: create a special TrimRight function that can trim to a certain length.
+        If(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:228]: we don't need 2 `If` if we don't need to trim.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala [hunk 2]
```diff
@@ -21,7 +21,7 @@ import scala.collection.mutable
 
 import org.apache.spark.internal.Logging
 import org.apache.spark.sql.AnalysisException
-import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.{StringTrimRight, _}
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:24]: why change the import?

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala (hunk 2 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:24]: Not me, the IDE did it:)

**Comment by dongjoon-hyun** [comment]
cc @HyukjinKwon since this affects 3.1.1.

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala [hunk 3]
```diff
@@ -666,6 +682,19 @@ class FileSourceCharVarcharTestSuite extends CharVarcharTestSuite with SharedSpa
       assert(rest.contains("CHAR(5)"))
     }
   }
+
+  test("SPARK-34114: should not trim right for read-side length check and char padding") {
+    Seq("char", "varchar").foreach { typ =>
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:687]: `typ` -> `type`?

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:687]: `type` is a scala keyword and can't be used as a variable name.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala [hunk 4]
```diff
@@ -208,53 +209,60 @@ object CharVarcharUtils extends Logging {
...
-          StringRPad(trimmed, Literal(length))))
+          raiseError("char", length),
+          trimmed)
+
+      case VarcharType(length) =>
+        if (needTrim) {
+          val trimmed = StringTrimRight(expr)
+          // Trailing spaces do not count in the length check. We need to retain the trailing spaces
+          // (truncate to length N), as there is no read-time padding for varchar type.
+          // TODO: create a special TrimRight function that can trim to a certain length.
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:228]: Do we have a JIRA for this TODO? Shall we make this as `IDed` TODO like `TODO(SPARK-XXX)`?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:228]: In general, yes.

We may keep improving here. @cloud-fan and I have considered more than 3 ways to simplify the logic here for performance improvement.

This TODO may be invalid now, How about just rm it? WDYT? @cloud-fan @dongjoon-hyun

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:228]: let's keep it as it's not added by this PR: https://github.com/apache/spark/pull/31181/files#diff-16753285e80505e04c445ea8ccee1dde7ae601ed7bae224e212d90d395f57928L225

**Comment by cloud-fan** [comment]
thanks, merging to master/3.1!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala (hunk 4 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:228]: k that's fine. but let's add an ID-ed todo in general. That's the standard approach to make sure they are on track.

**Pull request closed without merging**